src/stdatamodels/model_base.py

```diff
@@ -624,6 +624,13 @@ def to_fits(self, init, *args, **kwargs):
             with warnings.catch_warnings():
                 warnings.filterwarnings('ignore', message='Card is too long')
                 if self._no_asdf_extension:
+                    # For some old files that were written out before the
+                    # _no_asdf_extension existed, these will have an ASDF
+                    # extension, which may get passed along through extra_fits.
+                    # Avoid this.
+                    if "ASDF" in [hdu.name for hdu in ff._hdulist]:
+                        del ff._hdulist["ASDF"]
+                    # Use HDUList.writeto instead of AsdfInFits.write_to
                     ff._hdulist.writeto(init, *args, **kwargs)
```

Comment:
I wonder why we're even using AsdfInFits in this case.  I can't think of anything valuable it does since we're discarding the tree anyway.

Reply:
Do we keep around a reference to the HDUList in the in-memory datamodel?  I couldn't find it.  I think we're mostly using `AsdfInFIts` as structure to map the tree to FITS keywords on `save`, and then discarding the tree by using the `HDUList.writeto()` method.  So it does serve some purpose, though I don't like they we're poking around in private attributes.

Reply:
AsdfInFits doesn't know how to map to FITS keywords though, that's an stdatamodels feature.  Maybe we ought to change the stdatamodels.fits_support.to_fits method to return an HDUList, and construct the AsdfInFits in this method if we need it?

Reply:
That would make more sense, wouldn't it?  Let me have a look.

Reply:
It would have to return both the tree and the hdulist, as the tree has been modified in `fits_support.to_fits` function.

Reply:
...unless it is modifying the tree in-place.

Reply:
It doesn't modify the tree in-place, so the return signature will have to be a tuple of the hdulist and the new tree.